database/rrdcontext.c

```diff
@@ -0,0 +1,2125 @@
+    };
+    uuid_copy(tri.uuid, sc->chart_id);
+
+    RRDINSTANCE_ACQUIRED *ria = (RRDINSTANCE_ACQUIRED *)dictionary_set_and_acquire_item(rc->rrdinstances, sc->id, &tri, sizeof(tri));
+    RRDINSTANCE *ri = rrdinstance_acquired_value(ria);
+
+    ctx_get_dimension_list(&ri->uuid, rrdinstance_load_dimension, ri);
+    ctx_get_label_list(&ri->uuid, rrdinstance_load_clabel, ri);
+
+    rrdinstance_release(ria);
+    rrdcontext_release(rca);
+}
+
+static void rrdcontext_load_context_callback(VERSIONED_CONTEXT_DATA *ctx_data, void *data) {
+    RRDHOST *host = data;
+    (void)host;
+
+    RRDCONTEXT tmp = {
+        .id = string_strdupz(ctx_data->id),
+
+        // no need to set more data here
+        // we only need the hub data
+
+        .hub = *ctx_data,
+    };
+    dictionary_set((DICTIONARY *)host->rrdctx, string2str(tmp.id), &tmp, sizeof(tmp));
+}
+
+void rrdhost_load_rrdcontext_data(RRDHOST *host) {
+    if(unlikely(rrdcontext_enabled == CONFIG_BOOLEAN_NO))
+        return;
+
+    if(host->rrdctx) return;
+
+    rrdhost_create_rrdcontexts(host);
+    ctx_get_context_list(&host->host_uuid, rrdcontext_load_context_callback, host);
+    ctx_get_chart_list(&host->host_uuid, rrdinstance_load_chart_callback, host);
+}
+
+// ----------------------------------------------------------------------------
+// the worker thread
+
+static inline usec_t rrdcontext_queued_dispatch_ut(RRDCONTEXT *rc, usec_t now_ut) {
+
+    if(likely(rc->last_delay_calc_ut >= rc->last_queued_ut))
+        return rc->scheduled_dispatch_ut;
+
+    RRD_FLAGS flags = rc->last_queued_flags;
+
+    usec_t delay = LONG_MAX;
+    int i;
+    struct rrdcontext_reason *reason;
+    for(i = 0, reason = &rrdcontext_reasons[i]; reason->name ; reason = &rrdcontext_reasons[++i]) {
+        if(unlikely(flags & reason->flag)) {
+            if(reason->delay_ut < delay)
+                delay = reason->delay_ut;
+        }
+    }
+
+    if(unlikely(delay == LONG_MAX)) {
+        internal_error(true, "RRDCONTEXT: '%s', cannot find minimum delay of flags %x", string2str(rc->id), (unsigned int)flags);
+        delay = 60 * USEC_PER_SEC;
+    }
+
+    rc->last_delay_calc_ut = now_ut;
+    usec_t dispatch_ut = rc->scheduled_dispatch_ut = rc->last_queued_ut + delay;
+    return dispatch_ut;
+}
+
+#define MESSAGES_TO_SEND_PER_HOST 1000
+
+#define RRDCONTEXT_DELAY_AFTER_DB_ROTATION_SECS 120
+// #define RRDCONTEXT_CLEANUP_DELETED_EVERY_SECS 3600 // not needed, now it is done at dequeue
+#define RRDCONTEXT_HEARTBEAT_SECS 1
+
+#define WORKER_JOB_HOSTS            1
+#define WORKER_JOB_CHECK            2
+#define WORKER_JOB_SEND             3
+#define WORKER_JOB_DEQUEUE          4
+#define WORKER_JOB_RETENTION        5
+#define WORKER_JOB_QUEUED           6
+#define WORKER_JOB_CLEANUP          7
+#define WORKER_JOB_CLEANUP_DELETE   8
+
+usec_t rrdcontext_last_cleanup_ut = 0;
+usec_t rrdcontext_last_db_rotation_ut = 0;
+
+void rrdcontext_db_rotation(void) {
```

Comment:
A suggestion, this function could be an inline function.

Reply:
This is used from another file. It is not static.